### ISSUE TITLE
MathJax style parsing

### ISSUE THREAD

**Issue by mcmcclur** [discourse/discourse-math#7]
As much as I like this plugin, I do think the parsing could be much more natural. This was mentioned in several posts on [this meta thread](https://meta.discourse.org/t/discourse-math-plugin/65770). After playing with [markdown-it-mathjax](https://github.com/classeur/markdown-it-mathjax), I decided to see if I could incorporate that almost verbatim into this plugin. The result is, in my opinion an improvement. I've got the implementation in [this fork of discourse-math](https://github.com/mcmcclur/discourse-math). 

I'd love to do a pull request, but I'm pretty sure you'll want to take a good look first. To help, the rest of this post presents a comparison of the two parsers. For the purposes of this post, we'll call these two closely related dialects *Saffronese* and *Jaxian*.

 * Saffronese is the parser first implemented for the discourse-math plugin by the original author Sam Saffron.
 * Jaxian is implemented in [this fork](https://github.com/mcmcclur/discourse-math) of the discourse-math plugin. I take little credit for the code, though; it's really a fairly direct incorporation of the [markdown-it-mathjax plugin](https://github.com/classeur/markdown-it-mathjax).

The only relevant file for both the original plugin and the fork is

    assets/javascripts/lib/discourse-markdown/discourse-math.js.es6

In the image below, the left side presents the current Saffronese and the right side presents my proposed Jaxian. The markdown that produced both is below the image.

---

![out](https://user-images.githubusercontent.com/11695301/28990267-c22e753e-7949-11e7-9895-c81d22394e7e.png)

---

    ## Stuff that works fine in both

    Generally, inline math is very similar in both dialects. For example, $f(x) = e^{-x^2}$ implies $f'(x)=-2xe^{-x^2}$. Note that the first input has space on both sides, while the second input has a punctuation mark after it.

    The way that display math works in the two parsers is a bit different. Saffronese expects that the double dollar signs appear on separate newlines with only space following. Thus, Saffronese expects input like

        $$
        \int_{-\infty}^{\infty} f(x) = \sqrt{x}
        $$

    To produce

    $$
    \int_{-\infty}^{\infty} f(x) = \sqrt{x}
    $$

    Note that this type of input works in the new parser as well.

    One major enhancement present in Saffronese also works in Jaxian - namely the parser can tell when dollar signs are used to indicate money. For example:

    Well, I have $200 and Adrienne has $300.  Looks good!

    While we're still on the topic of things that work in both parsers, let's check that \begin{env}...\end{env}` type environments work when wrapped in double dollar signs. For example:

    $$
    \begin{align}
      f(x) &= x^2 - 3x + 2 \\
            &= (x-2)(x-1).
    \end{align}
    $$

    Good again!

    ## Jaxian advantages

    I think Jaxian has two major advantages. First, double dollar signs to indicate display math need not appear on new lines. That is, we can type $$\int_{-\infty}^{\infty} f(x) = \sqrt{x}$$ like so.

    Another nice item is that we can type `$x$s` plural - as in, the expression $x+x$ has two $x$s.  I wasn't able to get this to work in Saffronese at all.

    ## Open math

    Finally, there's one thing about Saffronese that I don't think that I like - namely, "unclosed `$$` is considered math". Thus, the following parses in his original parser but not in my new, suggested parser:

    $$
    \int_{-\infty}^{\infty} e^{-x^2} = \sqrt{x}

    I personally, think that's a bad idea.

**Comment by SamSaffron** [comment]
Sorry for the huge delay here, I was on holidays. 

I kept "open math" just to match open ``` , I thought parity with the way CommonMark works for source code makes sense. http://spec.commonmark.org/dingus/?text=%60%60%60%0Ax%2B%2B%3B%0A

Totally fine to add support for $this_is_math$s I guess it is ok to just skip boundary check on the closing dollar. 

Also totally fine to add support for $$x+x$$ , though I am not super happy that it is mixing inline with block level elements ... I would prefer if it was just treated exactly the same as $x+x$

**Comment by SamSaffron** [comment]
@mcmcclur any comments here, I definitely want to make the plugin easier for you!

**Comment by mcmcclur** [comment]
Yes, I definitely have some comments - though, I'm afraid I might come off as just reiterating things I already said. Honestly, the parser that I proposed in my original post (and called Jaxian, for some reason) works pretty much exactly as I would hope. Since you ask, I guess I should start by stating the three main things that I like better about it.
  1. I think that input like two $x$s should work,
  2. I think that display math on one line like
$$f(x) = e^x$$
should work, and
  3. I don't like the open $$

Let me try to support these points but first, I should say that, having worked on the parser a bit, I'm now very impressed with the approach you've taken to let markdown-it parse TeX to a span or div of class math. I also now realize that there's bound to be some differences between math input into a web box and math input into a LaTeX text editor.

*But* - a line like 
$$f(x) = e^x$$
with the double dollar signs on the same line is an extremely common idiom in LaTeX. If it's not particularly hard to allow that, I think it's a mistake not too. On the other hand, I don't particularly think it's necessary to allow display math to be typed inline like $$f(x)=e^x$$ but I don't think it hurts either.

As far as the open double $$ signs goes, that's an interesting point you make about the open \``` for code. But it still doesn't feel right - just try it. When you type a double $$, move to the next line, and keep typing, the MathJax processing starts right away. Thus, the whole thing re-typesets with every key stroke. It's quite annoying. That's not an issue with the triple \```.

And sometimes, you just need two $x$s! :)

So, I guess the guiding principle is - if it's not particularly difficult to incorporate a common LaTeX idiom then it should probably be done.

I spent some time earlier this week fiddling with your parser but wasn't quite able to get it working the way I'd like. Honestly, the mechanism that decides when something like an `md.block.ruler` actually fires is a mystery to me and I found the original code by classeur to be a bit easier to understand. 

I've actually put my small modification of his code into production for a class that starts next week. Unfortunately, that class is an elementary stats class and we're not likely to put it to the test to the degree that many of my classes do. I'll definitely be working with math in discourse for some time to come, though!

**Comment by mcmcclur** [comment]
Would you consider it if I just submit a pull-request for my code? I thought maybe not, which is why I created this issue, but I'm not really sure how else to approach it at this point.

**Comment by SamSaffron** [comment]
Split it up into multiple PRs. the $$ inline support is something that we
got to add
